ghcide/src/Development/IDE/Core/Rules.hs

```diff
@@ -827,6 +908,9 @@ getModIfaceWithoutLinkableRule = defineEarlyCutoff $ \GetModIfaceWithoutLinkable
       msg = "tried to look at linkable for GetModIfaceWithoutLinkable for " ++ show f
   pure (fingerprintToBS . getModuleHash . hirModIface <$> mhfr', ([],mhfr'))
 
+-- | Also generates and indexes the `.hie` file, along with the `.o` file if needed
+-- Invariant maintained is that if the `.hi` file was successfully written, then the
+-- `.hie` and `.o` file (if needed) were also successfully written
 regenerateHiFile :: HscEnvEq -> NormalizedFilePath -> ModSummary -> Maybe LinkableType -> Action ([FileDiagnostic], Maybe HiFileResult)
```

Comment:
Would it make sense to encode the indexing witness in the file system rather than rely on file write ordering?

Reply:
I don't see the advantage. What exactly do you have in mind?

Reply:
To write an empty file Foo.indexed as evidence that a module Foo has been indexed into the db

Reply:
That is not a problem we have over here - we can always consult the db directly to check that the correct file has been indexed (and verify it matches the file on disk by comparing the hash).

`.hie` files don't maintain any state for recompilation checking like `.hi` files, so it is not easy to tell if they are up to date (we can compare mtimes or the contents of the `.hs` file, but this doesn't capture dependencies changing, compilation options etc.). So the idea is to piggy back on `.hi` files for the correct behaviour. If `.hi` files need to be recompiled, we regenerate the `.hie` files too. However, if we write the `.hi` file before the `.hie` file, something can go wrong in between, so we end  up with an up2date `.hi` file, but an old `.hie` file, with no way to detect this situation. Changing the order of the writes should prevent this scenario.